wagtail/admin/compare.py

```diff
@@ -55,8 +57,155 @@ def htmldiff(self):
         ).to_html()
 
 
-class StreamFieldComparison(RichTextFieldComparison):
-    pass
+def get_comparison_class_for_block(block):
+    if hasattr(block, 'get_comparison_class'):
+        return block.get_comparison_class()
+    elif isinstance(block, blocks.CharBlock):
+        return CharBlockComparison
+    elif isinstance(block, blocks.RichTextBlock):
+        return RichTextBlockComparison
+    elif isinstance(block, blocks.StructBlock):
+        return StructBlockComparison
+    else:
+        # As all stream field blocks have a HTML representation, fall back to diffing that.
+        return RichTextBlockComparison
+
+
+class BlockComparison:
+    def __init__(self, block, val_a, val_b):
+        self.block = block
+        self.val_a = val_a
+        self.val_b = val_b
+
+    def is_new(self):
+        return bool(self.val_b) and not bool(self.val_a)
+
+    def is_deleted(self):
+        return bool(self.val_a) and not bool(self.val_b)
+
+    def has_changed(self):
+        return self.val_a != self.val_b
+
+
+class CharBlockComparison(BlockComparison):
+    def htmldiff(self):
+        return diff_text(
+            force_text(self.val_a),
+            force_text(self.val_b)
+        ).to_html()
+
+
+class RichTextBlockComparison(BlockComparison):
+    def htmldiff(self):
+        return diff_text(
+            BeautifulSoup(force_text(self.val_a), 'html5lib').getText(),
+            BeautifulSoup(force_text(self.val_b), 'html5lib').getText()
+        ).to_html()
+
+
+class StructBlockComparison(BlockComparison):
+    def htmldiff(self):
```

Comment:
The rendering of nested StructBlocks is a little bit inconsistent.

Consider these blocks:

```py
class PersonBlock(blocks.StructBlock):
    name = blocks.CharBlock()
    email = blocks.EmailBlock()

class TeamBlock(blocks.StructBlock):
    name = blocks.CharBlock()
    leader = PersonBlock()
```

Then say a page has two PersonBlocks and an editor adds a new TeamBlock:

![image](https://user-images.githubusercontent.com/654645/50705169-eb5cb080-1027-11e9-93f1-6a840571a233.png)

Unfortunately the diff doesn't indent the nested PersonBlock (seems like NAME and EMAIL should be indented a bit under LEADER):

![image](https://user-images.githubusercontent.com/654645/50705202-ffa0ad80-1027-11e9-8dfd-68a8bb021cde.png)

Additionally, let's say the PersonBlock has a custom template:


```py
class PersonBlock(blocks.StructBlock):
    name = blocks.CharBlock()
    email = blocks.EmailBlock()

    class Meta:
        template = 'blocks/person_block.html'
```

```html
<!-- blocks/person_block.html -->
<a href="mailto:{{ value.email }}">Email {{ value.name }}</a>
```

Now the nested block is still using its HTML representation (note the use of the custom template there that generates an email link), but the top-level block is using the basic rendering:

![image](https://user-images.githubusercontent.com/654645/50705392-b0a74800-1028-11e9-95f6-83eabcff180e.png)

This also unfortunately means that the bug mentioned on https://github.com/wagtail/wagtail/issues/3368#issuecomment-451496319 about custom block templates requiring template context can still cause errors when comparing.